File: cs/sdk/src/Proton.Drive.Sdk/Nodes/DegradedNode.cs
using System.Text.Json.Serialization;
using Proton.Sdk;

namespace Proton.Drive.Sdk.Nodes;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(DegradedFolderNode), typeDiscriminator: "folder")]
[JsonDerivedType(typeof(DegradedFileNode), typeDiscriminator: "file")]
[JsonDerivedType(typeof(DegradedPhotoNode), typeDiscriminator: "photo")]
public abstract record DegradedNode
{
    public required NodeUid Uid { get; init; }

    public required NodeUid? ParentUid { get; init; }

    public string TreeEventScopeId => Uid.VolumeId.ToString();

    public required Result<string, ProtonDriveError> Name { get; init; }

    public required Result<Author, SignatureVerificationError> NameAuthor { get; init; }

    public required DateTime CreationTime { get; init; }

    public DateTime? TrashTime { get; init; }

    public required Result<Author, SignatureVerificationError> Author { get; init; }

    public required IReadOnlyList<ProtonDriveError> Errors { get; init; }
}

File: kt/sdk/src/main/kotlin/me/proton/drive/sdk/entity/DegradedNode.kt
package me.proton.drive.sdk.entity

import java.time.Instant

sealed interface DegradedNode {
    val uid: NodeUid
    val parentUid: ParentNodeUid?
    val treeEventScopeId: ScopeId
    val name: Result<String>
    val creationTime: Instant
    val trashTime: Instant?
    val nameAuthor: Result<Author>
    val author: Result<Author>
    val errors: List<DriveError>
}
